### PR TITLE
feature(lxl): add infoNCE loss

### PR DESCRIPTION
**Pull request by lixl-st**
## Description
1. add infoNCE loss
2. add deep infomax as an auxiliary loss to the DQN

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/utils/mi_estimator.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
+from typing import Union
```
Comment (PaParaZz1) [inline ding/utils/mi_estimator.py:1]: move this file to `torch_utils`

File: ding/utils/mi_estimator.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    """
+        The class for contrastive learning losses.
+        Only InfoNCE loss supported currently.
+        Code Reference: https://github.com/rdevon/DIM.
+        paper: https://arxiv.org/abs/1808.06670.
+    """
+
+    def __init__(
+            self,
+            x_dim: Union[int, SequenceType],
```
Comment (PaParaZz1) [inline ding/utils/mi_estimator.py:19]: it would be better to rename to `x_dim` or `x_size`

File: ding/utils/mi_estimator.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            y_dim: Union[int, SequenceType],
+            heads: SequenceType = [1, 1],
+            encode_shape: int = 64,
+            loss_type: str = "infonce",
```
Comment (PaParaZz1) [inline ding/utils/mi_estimator.py:23]: `infonce` can be more clear

File: ding/utils/mi_estimator.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        else:
+            hidden_size_list = [32, 64, 64, self._encode_shape * heads]
+            if obs[-1] >= 36:
+                encoder = ConvEncoder(obs, hidden_size_list)
+            else:
+                encoder = ConvEncoder(obs, hidden_size_list, kernel_size=[4, 3, 2], stride=[2, 1, 1])
+        return encoder
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor):
+        '''
```
Comment (PaParaZz1) [inline ding/utils/mi_estimator.py:63]: use """ rather than '''

File: dizoo/atari/config/serial/pong/pong_dqn_aux_config.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
+from easydict import EasyDict
+
+pong_dqn_aux_config = dict(
+    exp_name='pong_dqn_seed0',
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/pong/pong_dqn_aux_config.py:4]: rename `exp_name` related to `aux`

File: ding/utils/mi_estimator.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        # Masking is done by shifting the diagonal before exp.
+        u_neg = (n_mask * u_all) - (10. * (1 - n_mask))  # mask out "self" examples, all diagonals are set to -10.
+        u_neg = u_neg.reshape(N, N * x_heads, y_heads).unsqueeze(dim=1).expand(-1, x_heads, -1, -1)
+
+        # Since this is multiclass, we concat the positive along the class dimension before performing log softmax.
+        pred_lgt = torch.cat([u_pos, u_neg], dim=2)
+        pred_log = F.log_softmax(pred_lgt * self._temperature, dim=2)
+
+        # The positive score is the first element of the log softmax.
+        loss = -pred_log[:, :, 0].mean()
```
Comment (PaParaZz1) [inline ding/utils/mi_estimator.py:98]: you can replace it with `pred_log[..., 0]`

File: ding/utils/mi_estimator.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        Args:
+            x_dim: input dimensions for x, both the obs shape and the encoding shape are supported.
+            y_dim: input dimensions for y, both the obs shape and the encoding shape are supported.
+            heads: a list of 2 int elems, heads[0] for x and head[1] for y.
+                Used in multi-head, global-local, local-local MI maximization process.
+            loss_type: only InfoNCE loss supported currently.
+            temperature: the parameter to adjust the log_softmax.
+        """
+        super(ContrastiveLoss, self).__init__()
+        assert len(heads) == 2
```
Comment (PaParaZz1) [inline ding/utils/mi_estimator.py:36]: assert len(heads) == 2, len(heads)

File: ding/utils/mi_estimator.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        Args:
+            x: the input x, both raw obs and encoding are supported.
+            y: the input y, both raw obs and encoding are supported.
+        Returns:
+            torch.Tensor: loss value.
+        '''
+
+        N = x.size(0)
+        x_heads, y_heads = self._heads
+        x = self._x_encoder.forward(x).reshape(N, x_heads, self._encode_shape)
```
Comment (PaParaZz1) [inline ding/utils/mi_estimator.py:74]: use `view` rather than `reshape` as much as possible, which the latter will rearrange memory

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/atari/config/serial/pong/pong_dqn_stdim_config.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+)
+pong_dqn_stdim_config = EasyDict(pong_dqn_stdim_config)
+main_config = pong_dqn_stdim_config
+pong_dqn_stdim_create_config = dict(
+    env=dict(
+        type='atari',
+        import_names=['dizoo.atari.envs.atari_env'],
+    ),
+    env_manager=dict(type='subprocess'),
+    policy=dict(type='dqn'),
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/pong/pong_dqn_stdim_config.py:57]: why still `dqn` here

File: ding/policy/dqn.py [hunk 10]
```diff
@@ -381,3 +382,235 @@ def default_model(self) -> Tuple[str, List[str]]:
...
+        self._target_model.update(self._learn_model.state_dict())
+        return {
+            'cur_lr': self._optimizer.defaults['lr'],
+            'total_loss': loss.item(),
+            'aux_loss': aux_loss_eval.item(),
+            'q_value': q_value.mean().item(),
+            'priority': td_error_per_sample.abs().tolist(),
+            # Only discrete action satisfying len(data['action'])==1 can return this and draw histogram on tensorboard.
+            # '[histogram]action_distribution': data['action'],
+        }
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:616]: add `aux_loss` in `monitor_var_learn` method so as to record this variable in tensorboard

File: ding/policy/dqn.py [hunk 11]
```diff
@@ -381,3 +382,235 @@ def default_model(self) -> Tuple[str, List[str]]:
...
+                # (float) Epsilon start value
+                start=0.95,
+                # (float) Epsilon end value
+                end=0.1,
+                # (int) Decay length(env step)
+                decay=10000,
+            ),
+            replay_buffer=dict(replay_buffer_size=10000, ),
+        ),
+        loss_ratio=0.01,
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:502]: unify this variable name, such as `aux_loss_ratio`, add config comment

File: ding/policy/dqn.py [hunk 12]
```diff
@@ -381,3 +382,235 @@ def default_model(self) -> Tuple[str, List[str]]:
...
+        super()._init_learn()
+        self._main_encoder = self._model.encoder
+        x_size, y_size = self._get_encoding_size()
+        self._aux_model = ContrastiveLoss(x_size, y_size, **self._cfg.aux_model)
+        if self._cuda:
+            self._aux_model.cuda()
+        self._aux_optimizer = Adam(self._aux_model.parameters(), lr=self._cfg.learn.learning_rate)
+        self._aux_ratio = self._cfg.loss_ratio
+
+    def _get_encoding_size(self):
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:515]: move this function to encoder class, and add assertion in policy

File: ding/policy/dqn.py [hunk 13]
```diff
@@ -381,3 +382,235 @@ def default_model(self) -> Tuple[str, List[str]]:
...
+    )
+
+    def _init_learn(self) -> None:
+        super()._init_learn()
+        self._main_encoder = self._model.encoder
+        x_size, y_size = self._get_encoding_size()
+        self._aux_model = ContrastiveLoss(x_size, y_size, **self._cfg.aux_model)
+        if self._cuda:
+            self._aux_model.cuda()
+        self._aux_optimizer = Adam(self._aux_model.parameters(), lr=self._cfg.learn.learning_rate)
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:512]: you should overwrite `state_dict` and `load_state_dict` method with `aux_optimizer`

File: ding/policy/dqn.py [hunk 14]
```diff
@@ -10,6 +10,7 @@
 from ding.utils.data import default_collate, default_decollate
 from .base_policy import Policy
 from .common_utils import default_preprocess_learn
+from ding.torch_utils import ContrastiveLoss
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:13]: polish import sequence

File: ding/policy/dqn.py [hunk 15]
```diff
@@ -381,3 +382,235 @@ def default_model(self) -> Tuple[str, List[str]]:
...
+                                                                                                 | value to end value
+                                                                                                 | during decay length.
+        20 | ``loss_ratio``     float    0.01           | the ratio of auxiliary loss to main    | any real value,
+                                                        | loss                                   | typically in
+                                                                                                 | [-0.1, 0.1].
+        == ==================== ======== ============== ======================================== =======================
+    """
+
+    config = dict(
+        type='dqn',
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:447]: modify type

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — APPROVED** [review]
